File: ds/queue.py
class Node:
    def __init__(self, value, next=None) -> None:
        self.value = value
        self.next = next


class Queue:
    def __init__(self) -> None:
        self.head = None
        self.tail = None
        self.n = 0

    def offer(self, value):
        """Add an element to the end of the queue"""
        node = Node(value)
        self.n += 1
        if self.tail is None:
            self.head = self.tail = node

        self.tail.next = node
        self.tail = node

    def peek(self):
        """Look at the head element of the queue without removing it"""
        if self.head is None:
            return None
        return self.head.value

    def poll(self):
        """Remove the head element and return the objects value"""
        if self.head is None:
            return None
        tmp = self.head
        self.head = self.head.next
        self.n -= 1
        return tmp.value

    def size(self):
        return self.n

File: go/structures/queue.go
package structures

import "errors"

// Queue is an interface representing a First-In-First-Out (FIFO) data structure.
type Queue[T any] interface {
	// Enqueue adds an element to the rear of the queue.
	Enqueue(item T)
	// Dequeue removes and returns the front element of the queue.
	// Returns an error if the queue is empty.
	Dequeue() (T, error)
	// Peek returns the front element of the queue without removing it.
	// Returns an error if the queue is empty.
	Peek() (T, error)
	// IsEmpty returns true if the queue is empty, false otherwise.
	IsEmpty() bool
	// Size returns the number of elements in the queue.
	Size() int
	// Clear removes all elements from the queue.
	Clear()
}

var ErrEmptyQueue = errors.New("queue is empty")

type ListQueue[T any] struct {
	head *AnyNode[T]
	tail *AnyNode[T]
	size int
}

func (q *ListQueue[T]) Enqueue(item T) {
	q.size++
	node := &AnyNode[T]{value: item}
	if q.tail == nil {
		q.head = node
		q.tail = node
		return
	}

	q.tail.next = node
	q.tail = node
}

func (q *ListQueue[T]) Dequeue() (T, error) {
	if q.head == nil {
		var zero T
		return zero, ErrEmptyQueue
	}

	q.size--
	current := q.head
	q.head = q.head.next

	if q.head == nil {
		q.tail = nil
	}
	return current.value, nil
}

func (q *ListQueue[T]) Peek() (T, error) {
	if q.head == nil {
		var zero T
		return zero, ErrEmptyQueue
	}
	return q.head.value, nil
}

func (q *ListQueue[T]) IsEmpty() bool {
	return q.head == nil
}

func (q *ListQueue[T]) Size() int {
	return q.size
}

func (q *ListQueue[T]) Clear() {
	q.size = 0
	q.head = nil
	q.tail = nil
}
